### PR TITLE
[p2p] Pong Timeout

### PR DESCRIPTION
**Pull request by melekes**
Refs #804

Original PR: https://github.com/tendermint/tendermint/pull/832

### CODE REVIEW

**Comment by melekes** [comment]
2 peers
4 channels

```
BenchmarkSwitchBroadcast-2         30000             71275 ns/op
--- BENCH: BenchmarkSwitchBroadcast-2
        switch_test.go:339: success: 1, failure: 0
        switch_test.go:339: success: 100, failure: 0
        switch_test.go:339: success: 10000, failure: 0
        switch_test.go:339: success: 30000, failure: 0
```

Tested code:

```
func (sw *Switch) Broadcast(chID byte, msg interface{}) chan bool {
	successChan := make(chan bool, len(sw.peers.List()))
	sw.Logger.Debug("Broadcast", "channel", chID, "msg", msg)
	var wg sync.WaitGroup
	for _, peer := range sw.peers.List() {
		wg.Add(1)
		go func(peer Peer) {
			defer wg.Done()
			success := peer.Send(chID, msg)
			successChan <- success
		}(peer)
	}
	go func() {
		wg.Wait()
		close(successChan)
	}()
	return successChan
}
```

File: p2p/connection.go [hunk 1]
```diff
@@ -176,7 +186,11 @@ func (c *MConnection) OnStart() error {
 	}
 	c.quit = make(chan struct{})
 	c.flushTimer = cmn.NewThrottleTimer("flush", c.config.flushThrottle)
-	c.pingTimer = cmn.NewRepeatTimer("ping", pingTimeout)
+	c.pingTimer = cmn.NewRepeatTimer("ping", c.config.pingInterval)
+	c.pongTimer = time.NewTimer(c.config.pongTimeout)
+	// we start timer once we've send ping; needed here because we use start
+	// listening in recvRoutine
+	_ = c.pongTimer.Stop()
```
Comment (caffix) [inline p2p/connection.go]: Why the assignment to only a blank identifier?

File: p2p/connection.go [hunk 2]
```diff
@@ -188,6 +202,7 @@ func (c *MConnection) OnStop() {
 	c.BaseService.OnStop()
 	c.flushTimer.Stop()
 	c.pingTimer.Stop()
+	_ = c.pongTimer.Stop()
```
Comment (caffix) [inline p2p/connection.go]: Same as the question/comment above related to the blank identifier.

File: p2p/switch.go [hunk 3]
```diff
@@ -398,16 +399,22 @@ func (sw *Switch) IsDialing(addr *NetAddress) bool {
 // trying to send for defaultSendTimeoutSeconds. Returns a channel
 // which receives success values for each attempted send (false if times out).
 // NOTE: Broadcast uses goroutines, so order of broadcast may not be preserved.
-// TODO: Something more intelligent.
 func (sw *Switch) Broadcast(chID byte, msg interface{}) chan bool {
```
Comment (caffix) [inline p2p/switch.go]: Regarding issue #805: What if we kept the implementation similar to this, yet each broadcast was queued, so that broadcast X must finish before broadcast Y? The individual messages could be sent via go-routines, but the order of broadcasts would be preserved. Thoughts?

File: p2p/connection.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ebuchman) [reply p2p/connection.go]: probably for sake of the linter. lets just check the value and return error if its false ?

[the remark this answers is not in the document]
Comment (melekes) [reply p2p/connection.go]: I used it to indicate that we don't care about return value here. but I can add `// nolint`

File: p2p/connection.go [hunk 4]
```diff
@@ -456,8 +477,10 @@ FOR_LOOP:
 			c.Logger.Debug("Receive Ping")
 			c.pong <- struct{}{}
 		case packetTypePong:
```
[the remark this answers is not in the document]
Comment (melekes) [reply p2p/connection.go]: nice idea

File: p2p/switch.go (hunk 3 above)
Comment (melekes) [reply p2p/switch.go]: Even better to have a queue for each peer. That way we won't be waiting for some peer!

**Comment by melekes** [comment]
I will handle broadcast in a separate PR.

File: p2p/connection.go [hunk 5]
```diff
@@ -319,6 +334,12 @@ FOR_LOOP:
 			legacy.WriteOctet(packetTypePing, c.bufWriter, &n, &err)
 			c.sendMonitor.Update(int(n))
 			c.flush()
+			c.Logger.Debug("Starting pong timer")
+			c.pongTimer.Reset(c.config.pongTimeout)
+		case <-c.pongTimer.C:
+			c.Logger.Debug("Pong timeout")
+			// XXX: should we decrease peer score instead of closing connection?
```
[the remark this answers is not in the document]
Comment (zramsay) [reply p2p/connection.go]: todo's should be issues, linked in code

**Comment by ebuchman** [comment]
Thanks Anton!

Do you think we could get this rebased on the `addrbook` branch (https://github.com/tendermint/tendermint/pull/1129) ? Alternatively we could try to merge this into that branch - whatever you think will be easier

File: p2p/conn/connection.go [hunk 6]
```diff
@@ -319,6 +341,12 @@ FOR_LOOP:
 			legacy.WriteOctet(packetTypePing, c.bufWriter, &n, &err)
 			c.sendMonitor.Update(int(n))
 			c.flush()
+			c.Logger.Debug("Starting pong timer")
+			c.pongTimer.Reset(c.config.PongTimeout)
+		case <-c.pongTimer.C:
+			c.Logger.Debug("Pong timeout")
+			// XXX: should we decrease peer score instead of closing connection?
```
Comment (melekes) [inline p2p/conn/connection.go]: ?

Comment (ebuchman) [reply p2p/conn/connection.go]: no scoring for now, lets just close the conn

File: p2p/conn/connection.go [hunk 7]
```diff
@@ -172,7 +189,11 @@ func (c *MConnection) OnStart() error {
 	}
 	c.quit = make(chan struct{})
 	c.flushTimer = cmn.NewThrottleTimer("flush", c.config.FlushThrottle)
-	c.pingTimer = cmn.NewRepeatTimer("ping", pingTimeout)
+	c.pingTimer = cmn.NewRepeatTimer("ping", c.config.PingInterval)
+	c.pongTimer = time.NewTimer(c.config.PongTimeout)
+	// we start timer once we've send ping; needed here because we use start
+	// listening in recvRoutine
+	_ = c.pongTimer.Stop()
```
Comment (ebuchman) [inline p2p/conn/connection.go]: seems bad - what if the timer fires before we call stop?  maybe we can just create it the first time we need it ?  Otherwise we probably need the 

```
if !t.Stop() {
	<-t.C
}
```

here

File: p2p/conn/connection.go [hunk 8]
```diff
@@ -452,8 +480,10 @@ FOR_LOOP:
 			c.Logger.Debug("Receive Ping")
 			c.pong <- struct{}{}
 		case packetTypePong:
-			// do nothing
 			c.Logger.Debug("Receive Pong")
+			if !c.pongTimer.Stop() {
```
Comment (ebuchman) [inline p2p/conn/connection.go]: what if we receive multiple pongs in a row ? wont we stop on the first one, and then this will return false and we'll block forever ?

File: p2p/conn/connection_test.go [hunk 9]
```diff
@@ -116,6 +119,40 @@ func TestMConnectionStatus(t *testing.T) {
 	assert.Zero(status.Channels[0].SendQueueSize)
 }
 
+func TestMConnectionPongTimeoutResultsInError(t *testing.T) {
```
Comment (ebuchman) [inline p2p/conn/connection_test.go]: Think we need more tests:

- ping/pong works correctly (ie. we don't disconnect)
- receiving many pongs is benign
- pong timeout results in error

File: p2p/conn/connection.go (hunk 6 above)
Comment (ebuchman) [reply p2p/conn/connection.go]: lets just remove the comment

File: p2p/conn/connection.go (hunk 7 above)
Comment (melekes) [reply p2p/conn/connection.go]: hmm. it could if pongTimeout is 0 or 1 ms... but it's 45 sec by default

Comment (ebuchman) [reply p2p/conn/connection.go]: but what about in tests when its ms ? could be a source of non-deterministic failures :(

Comment (melekes) [reply p2p/conn/connection.go]: sure, you're right. I'll fix. sorry

**Comment by melekes** [comment]
@ebuchman added 2 more tests

File: p2p/conn/connection.go [hunk 10]
```diff
@@ -318,13 +340,23 @@ FOR_LOOP:
 			c.Logger.Debug("Send Ping")
 			legacy.WriteOctet(packetTypePing, c.bufWriter, &n, &err)
 			c.sendMonitor.Update(int(n))
+			c.Logger.Debug("Starting pong timer", "dur", c.config.PongTimeout)
+			c.pongTimer = time.AfterFunc(c.config.PongTimeout, func() {
+				c.pongTimeoutCh <- struct{}{}
```
Comment (ebuchman) [inline p2p/conn/connection.go]: Cool! Will this leak ? If the conn stops before we receive this, will there be a stuck go-routine? Maybe we need to drain this channel when we call pongTimer.Stop ?

File: p2p/conn/connection.go [hunk 11]
```diff
@@ -452,8 +484,10 @@ FOR_LOOP:
 			c.Logger.Debug("Receive Ping")
 			c.pong <- struct{}{}
```
Comment (ebuchman) [inline p2p/conn/connection.go]: This pong channel is unbuffered. I think it means we could block here if the sendRoutine exits. This might be reason why StopPeerForError doesn't work sometimes? Might also play into a leak, Would you mind thinking through that and let me know if you agree or if I'm wrong for some reason ?

File: p2p/conn/connection_test.go [hunk 12]
```diff
@@ -116,6 +119,139 @@ func TestMConnectionStatus(t *testing.T) {
...
+		// read ping
+		server.Read(make([]byte, 1))
+		serverGotPing <- struct{}{}
+		// respond with pong
+		_, err = server.Write([]byte{packetTypePong})
+		require.Nil(t, err)
+	}()
+	<-serverGotPing
+
+	pongTimerExpired := mconn.config.PongTimeout + 10*time.Millisecond
```
Comment (ebuchman) [inline p2p/conn/connection_test.go]: did this test fail with the old code ? 

my understanding was in this scenario with the previous code, the client should have been deadlocked after the multiple pongs

File: p2p/conn/connection.go [hunk 13]
```diff
@@ -452,8 +484,10 @@ FOR_LOOP:
...
 		case packetTypePong:
-			// do nothing
 			c.Logger.Debug("Receive Pong")
+			if c.pongTimer != nil {
```
Comment (ebuchman) [inline p2p/conn/connection.go]: won't this race with sendRoutine? If the ping timer fires around when we receive a pong

Comment (melekes) [reply p2p/conn/connection.go]: it probably will. what's the solution? mutex and another channel?

File: p2p/conn/connection.go (hunk 10 above)
Comment (melekes) [reply p2p/conn/connection.go]: hmm.. it will, most probably

File: p2p/conn/connection.go (hunk 13 above)
Comment (melekes) [reply p2p/conn/connection.go]: also interesting why go racer does not say anything...

File: p2p/conn/connection_test.go (hunk 12 above)
Comment (melekes) [reply p2p/conn/connection_test.go]: sending multiple pings, not pongs, yes!

File: p2p/conn/connection.go (hunk 11 above)
Comment (melekes) [reply p2p/conn/connection.go]: How it prevents StopPeerForError from working? if the code is inside a goroutine. Leak? yes

File: p2p/conn/connection.go (hunk 10 above)
Comment (melekes) [reply p2p/conn/connection.go]: https://github.com/tendermint/tendermint/pull/1095/commits/a67865cbd91b3dde2cd54690bcaf77beca6abd01

File: p2p/conn/connection.go [hunk 14]
```diff
@@ -315,13 +337,21 @@ FOR_LOOP:
...
+		case <-c.pongTimeoutCh:
+			c.Logger.Debug("Pong timeout")
+			err = errors.New("pong timeout")
 		case <-c.pong:
 			c.Logger.Debug("Send Pong")
 			wire.WriteByte(packetTypePong, c.bufWriter, &n, &err)
 			c.sendMonitor.Update(int(n))
 			c.flush()
 		case <-c.quit:
+			c.stopPongTimer()
```
Comment (ebuchman) [inline p2p/conn/connection.go]: if theres an error in this loop, we just break out without processing the closed quit channel so pong timer won't get stopped. but we can't put it in OnStop because it will race. So I think it needs to also be called in the error != nil block below

File: p2p/conn/connection.go [hunk 15]
```diff
@@ -446,15 +476,16 @@ FOR_LOOP:
...
 			c.Logger.Debug("Receive Ping")
 			select {
 			case c.pong <- struct{}{}:
 			case <-c.quit:
 				break FOR_LOOP
 			}
 		case packetTypePong:
-			// do nothing
 			c.Logger.Debug("Receive Pong")
+			c.stopPongTimer()
```
Comment (ebuchman) [inline p2p/conn/connection.go]: if we receive ping and pong at the same time, wont this race with the `c.pongTimer = ...` in the sendRoutine ? Maybe we need a test where that happens for the race detector to pick it up

File: p2p/conn/connection.go [hunk 16]
```diff
@@ -506,6 +537,15 @@ FOR_LOOP:
 	}
 }
 
+func (c *MConnection) stopPongTimer() {
+	if c.pongTimer != nil {
+		if !c.pongTimer.Stop() {
+			<-c.pongTimer.C
```
Comment (ebuchman) [inline p2p/conn/connection.go]: suppose we receive two pongs in a row. wont we block here on the second one ?

Comment (ebuchman) [reply p2p/conn/connection.go]: Here's a diff that proves the point. All I needed to do was add a sleep to the test (without the sleep the pongTimer was just nil which is why it was passing). With the sleep, the test hangs forever

```
diff --git a/p2p/conn/connection.go b/p2p/conn/connection.go
index 40f37d29..b05a36ea 100644
--- a/p2p/conn/connection.go
+++ b/p2p/conn/connection.go
@@ -484,6 +484,7 @@ FOR_LOOP:
                                break FOR_LOOP
                        }
                case packetTypePong:
+                       fmt.Println("GOT PONG")
                        c.Logger.Debug("Receive Pong")
                        c.stopPongTimer()
                case packetTypeMsg:
@@ -538,12 +539,21 @@ FOR_LOOP:
 }
 
 func (c *MConnection) stopPongTimer() {
+       fmt.Println("STOP PONG TIMER. Is nil", c.pongTimer == nil)
        if c.pongTimer != nil {
                if !c.pongTimer.Stop() {
+                       fmt.Println("waiting on .C")
                        <-c.pongTimer.C
+                       fmt.Println("done waiting")
+               } else {
+                       fmt.Println("STOP TIMER WORKED")
                }
+               fmt.Println("draining")
                drain(c.pongTimeoutCh)
+               fmt.Println("Drained")
        }
+
+       fmt.Println("-------------")
 }
 
 type ConnectionStatus struct {
diff --git a/p2p/conn/connection_test.go b/p2p/conn/connection_test.go
index 270b4ae9..6efc6e08 100644
--- a/p2p/conn/connection_test.go
+++ b/p2p/conn/connection_test.go
@@ -173,6 +173,7 @@ func TestMConnectionMultiplePongsInTheBeginning(t *testing.T) {
        err := mconn.Start()
        require.Nil(t, err)
        defer mconn.Stop()
+       time.Sleep(time.Second * 3)
 
        // sending 3 pongs in a row (abuse)
        _, err = server.Write([]byte{packetTypePong})
```

File: p2p/conn/connection.go (hunk 15 above)
Comment (ebuchman) [reply p2p/conn/connection.go]: Oh the race detector actually detects this for me

Comment (ebuchman) [reply p2p/conn/connection.go]: I guess we could just wrap pongTimer in a mutex but that doesn't really feel right.

An alternative is maybe to use a separate go-routine to handle a pongTimer. Then sendRoutine and recvRoutine can just talk to the pongTimerRoutine via channel/s when they need to start or stop the pongTimer.

File: p2p/conn/connection.go (hunk 14 above)
Comment (melekes) [reply p2p/conn/connection.go]: You're right. Thanks

File: p2p/conn/connection.go (hunk 15 above)
Comment (melekes) [reply p2p/conn/connection.go]: What if we use the same pongTimeoutCh to signal to cancel timeout? See my last commit.

File: p2p/conn/connection.go (hunk 16 above)
Comment (melekes) [reply p2p/conn/connection.go]: I don't think this diff is correct (proves anything) because if you add `time.Sleep`, `sendRoutine` will attempt to send ping and block on `flush`. And since we're using a pipe here and you're not reading the ping (see `server.Read` calls), this whole thing will block.

Comment (ebuchman) [reply p2p/conn/connection.go]: that could be. but do you see what I'm referring to ? It still seems like it will deadlock if it receives multiple pongs

File: p2p/conn/connection.go [hunk 17]
```diff
@@ -506,6 +545,17 @@ FOR_LOOP:
...
 }
 
+// not goroutine-safe
+func (c *MConnection) stopPongTimer() {
+	if c.pongTimer != nil {
+		if !c.pongTimer.Stop() {
+			<-c.pongTimer.C
+		}
+		drain(c.pongTimeoutCh)
+		c.pongTimer = nil
```
Comment (melekes) [inline p2p/conn/connection.go]: this should prevent a deadlock with multiple stops/pongs

**Comment by melekes** [comment]
Done

File: p2p/conn/connection.go [hunk 18]
```diff
@@ -179,12 +201,12 @@ func (c *MConnection) OnStart() error {
...
 	c.BaseService.OnStop()
-	c.flushTimer.Stop()
-	c.pingTimer.Stop()
-	c.chStatsTimer.Stop()
 	if c.quit != nil {
 		close(c.quit)
 	}
+	c.flushTimer.Stop()
+	c.pingTimer.Stop()
+	c.chStatsTimer.Stop()
```
Comment (jaekwon) [inline p2p/conn/connection.go]: reason for moving these?

File: p2p/conn/connection.go [hunk 19]
```diff
@@ -315,7 +337,18 @@ FOR_LOOP:
 			c.Logger.Debug("Send Ping")
 			wire.WriteByte(packetTypePing, c.bufWriter, &n, &err)
 			c.sendMonitor.Update(int(n))
+			c.Logger.Debug("Starting pong timer", "dur", c.config.PongTimeout)
+			c.pongTimer = time.AfterFunc(c.config.PongTimeout, func() {
+				c.pongTimeoutCh <- true
```
Comment (jaekwon) [inline p2p/conn/connection.go]: this could/should be non-blocking.  probably select on .quit

File: p2p/conn/connection.go [hunk 20]
```diff
@@ -503,6 +542,17 @@ FOR_LOOP:
 	}
 }
 
+// not goroutine-safe
+func (c *MConnection) stopPongTimer() {
+	if c.pongTimer != nil {
+		if !c.pongTimer.Stop() {
+			<-c.pongTimer.C
+		}
+		drain(c.pongTimeoutCh)
```
Comment (jaekwon) [inline p2p/conn/connection.go]: I am fairly sure that we are better off avoiding draining functions.
If you're not convinced, I could try to put more thought into it, but I suspect that draining is not necessary, and consistently avoiding it is better.

Here, we can keep pongTimeoutCh capacity 0, and select on line 342.

File: p2p/conn/connection.go [hunk 21]
```diff
@@ -447,15 +481,20 @@ FOR_LOOP:
...
 				// never block
 			}
 		case packetTypePong:
-			// do nothing
 			c.Logger.Debug("Receive Pong")
+			select {
+			case c.pongTimeoutCh <- false:
+			default:
+				// never block
+			}
```
Comment (jaekwon) [inline p2p/conn/connection.go]: good use of default

File: p2p/conn/connection.go (hunk 20 above)
Comment (jaekwon) [reply p2p/conn/connection.go]: It's OK to leave things in channels, as long as nothing blocks on it.  We can drain defensively, but lets instead make sending to pongTimeoutCh nonblocking, and not drain.

File: p2p/conn/connection.go [hunk 22]
```diff
@@ -315,7 +337,18 @@ FOR_LOOP:
 			c.Logger.Debug("Send Ping")
 			wire.WriteByte(packetTypePing, c.bufWriter, &n, &err)
 			c.sendMonitor.Update(int(n))
+			c.Logger.Debug("Starting pong timer", "dur", c.config.PongTimeout)
+			c.pongTimer = time.AfterFunc(c.config.PongTimeout, func() {
```
Comment (jaekwon) [inline p2p/conn/connection.go]: I think we need to stop the previous pongTimer, if it exists.

File: p2p/conn/connection.go [hunk 23]
```diff
@@ -81,10 +83,15 @@ type MConnection struct {
...
-	flushTimer   *cmn.ThrottleTimer // flush writes as necessary but throttled.
-	pingTimer    *cmn.RepeatTimer   // send pings periodically
-	chStatsTimer *cmn.RepeatTimer   // update channel stats periodically
+	quit       chan struct{}
+	flushTimer *cmn.ThrottleTimer // flush writes as necessary but throttled.
+	pingTimer  *cmn.RepeatTimer   // send pings periodically
+
+	// close conn if pong is not received in pongTimeout
+	pongTimer     *time.Timer
+	pongTimeoutCh chan bool // true - timeout, false - peer sent pong
```
Comment (jaekwon) [inline p2p/conn/connection.go]: seems like it would better be,
true ~ timeout, close ~ peer sent pong.

File: p2p/conn/connection.go [hunk 24]
```diff
@@ -315,7 +337,18 @@ FOR_LOOP:
...
+			c.pongTimer = time.AfterFunc(c.config.PongTimeout, func() {
+				c.pongTimeoutCh <- true
+			})
 			c.flush()
+		case timeout := <-c.pongTimeoutCh:
+			if timeout {
+				c.Logger.Debug("Pong timeout")
+				err = errors.New("pong timeout")
+			} else {
+				c.stopPongTimer()
```
Comment (jaekwon) [inline p2p/conn/connection.go]: weird that we stop it here... i don't think this is the right place to be stopping it.

Comment (jaekwon) [reply p2p/conn/connection.go]: also, see my first comment way up.

File: p2p/conn/connection.go (hunk 21 above)
Comment (jaekwon) [reply p2p/conn/connection.go]: I think

File: p2p/conn/connection.go (hunk 22 above)
Comment (melekes) [reply p2p/conn/connection.go]: 1) there is no (little) chance it exists since above I check for `if config.PongTimeout >= config.PingInterval {`
2) that would be bad. what's the reason to start it in the first place if later we stop it and replace with newer timer

File: p2p/conn/connection.go (hunk 24 above)
Comment (melekes) [reply p2p/conn/connection.go]: weird?) we need to stop it here to prevent pongTimer from firing (`c.pongTimeoutCh <- true` above). 

```
pongTimerStart ............. (got pong -> stop pongTimer) xxx (pongTimerFires) ..... (pingTimer fires)
pongTimerStart ............. (pongTimer fires)
```

File: p2p/conn/connection.go (hunk 22 above)
Comment (ebuchman) [reply p2p/conn/connection.go]: seems correct but does Go guarantee it ? can we check if the pongTimer is still active ? if so, then logically it's timed out (unless we have a bug and fail to stop it when we receive the pong)

File: p2p/conn/connection.go (hunk 19 above)
Comment (ebuchman) [reply p2p/conn/connection.go]: hmm - won't that cause us to miss a lot of these ? if we're sending a peer lots of data, but it never responds to pings, we might never detect that.

File: p2p/conn/connection.go (hunk 23 above)
Comment (ebuchman) [reply p2p/conn/connection.go]: but then you'd have to keep recreating the channel. why would that be better ?

File: p2p/conn/connection.go (hunk 20 above)
Comment (ebuchman) [reply p2p/conn/connection.go]: This is done in https://github.com/tendermint/tendermint/pull/1209, but is it really ok if we're always ignoring those timeouts ?
